### ISSUE TITLE
Issue during test with different batch sizes

### ISSUE THREAD

**Issue by aminpdik** [niejiahao1998/IFA#16]
Hello @niejiahao1998 

I am exploring your paper and the provided code. I encountered the same issue as #6 using your provided trained models.

I believe I’ve identified the cause of the problem. In line 75 of the test code, you have cls = cls[0].item() followed by cls = cls + 1. Later, in lines 82 and 83, you modify the binary prediction and mask using this cls value. The issue arises because the images in the batch belong to different classes, and the cls tensor has the class id of query images. However, you’re always using cls[0] (the class of the first image in the batch) for all images, even though they belong to other classes.

This means the correct accuracy can only be achieved when the batch size is set to 1, as there’s only one image and one class in the cls tensor in that case.

Could you please confirm if my understanding is correct?

**Comment by niejiahao1998** [comment]
Thanks for pointing it out. We carefully check the codes and please let me explain it.

1-Our code is adapted from SSP, thus cls[0] is written by previous repo.
2-The different performance from bsz=1 and bsz=6 is a bug, you are correct.
3-Acutually, we should make sure the bsz>#class. For example, deepglobe has 6 classes, thus we should make sure the bsz larger than 6. Otherwise, the 6th category will be neglected and thus we have a lower performance.
4-Please forgive me that it is really hard to explain because the logic between the original code from SSP is very complex. But please understand that what we need to do is make sure the bsz>#class.
5-You may check the code line by line, and hope you can understand what I am trying to explain... It is really weird and strange.

Sorry for this bug, and thanks again for your check.

**Comment by Sparkling-Water** [comment]
> Hello @niejiahao1998
> 
> I am exploring your paper and the provided code. I encountered the same issue as #6 using your provided trained models.
> 
> I believe I’ve identified the cause of the problem. In line 75 of the test code, you have cls = cls[0].item() followed by cls = cls + 1. Later, in lines 82 and 83, you modify the binary prediction and mask using this cls value. The issue arises because the images in the batch belong to different classes, and the cls tensor has the class id of query images. However, you’re always using cls[0] (the class of the first image in the batch) for all images, even though they belong to other classes.
> 
> This means the correct accuracy can only be achieved when the batch size is set to 1, as there’s only one image and one class in the cls tensor in that case.
> 
> Could you please confirm if my understanding is correct?

I totally agree with you, and I tested the trained model that the author provided using the mIoU calculation methods from both the IFA code(setting 1) and the PATNet code(setting 2). Here are the results:

	         Deepglobe	ISIC	    Chest X-Ray	FSS-1000	Average
setting 1	    50.23	       71.45	   74.51	           81.38	          69.39
setting 2	    44.21	       55.99	   74.51	           79.75	          63.61

According to the mIoU calculation method in the IFA code, when the batch size is greater than 1, some mask category IDs mixed up, leading to differences like (in_a/un_a) + (in_b/un_b) and (in_a + in_b) / (un_a + un_b).
For example, in the evaluation of the Deepglobe dataset, the hist is as follows: 

[[78917153. 26400543.        0.        0.        0.        0.]
 [35609113. 51073191.        0.        0.        0.        0.]
 [       0.        0.        0.        0.        0.        0.]
 [       0.        0.        0.        0.        0.        0.]
 [       0.        0.        0.        0.        0.        0.]
 [       0.        0.        0.        0.        0.        0.]]

only the first two classes calculated, while the values for the rows and cols corresponding to the other classes are all 0.

**Comment by HeShiLie** [comment]
hello, @niejiahao1998 , I check the source code, I think that the problem is not to set "bsz >= num_classes". In fact, if the codes in `class mIOU` is not wrong, I think the problem is what @aminpdik said, here is my repairment for such bug:
```python
        cls = cls + 1 # cls.shape: (b)
        cls = repeat(cls, 'b -> b h w', h=mask_q.shape[1], w=mask_q.shape[2]).to(device) # cls: (b, h, w)
        with torch.no_grad():
            pred = model(img_s_list, mask_s_list, img_q, None)[0]
            pred = torch.argmax(pred, dim=1)

        pred[pred == 1] = cls[pred == 1] # pred: (b, h, w)
        mask_q[mask_q == 1] = cls[mask_q == 1].to(dtype = mask_q.dtype) # mask_q: (b, h, w)
```
i test the new eval code for different bsz like 4,8,16,32 under the same seed and perfomance the the same. I think this is an efficient way for such bug.

**Comment by King-king424** [comment]
I think the best way is to use the same mIoU calculation code as PATNet.
